x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -140,96 +149,63 @@ public boolean release(final EvictionListener listener) {
         assert listener != null;
 
         boolean success = false;
-        evictionLock.lock();
         try {
-            try {
-                final Set<EvictionListener> newListeners = new HashSet<>(listeners);
-                final boolean removed = newListeners.remove(Objects.requireNonNull(listener));
+            synchronized (listeners) {
+                final boolean removed = listeners.remove(Objects.requireNonNull(listener));
                 assert removed : "listener does not exist " + listener;
                 if (removed == false) {
                     throw new IllegalStateException("Cannot remove an unknown listener");
                 }
-                maybeCloseFileChannel(newListeners);
-                listeners = Collections.unmodifiableSet(newListeners);
-                success = true;
-            } finally {
-                if (success) {
-                    refCounter.decRef();
+                if (listeners.isEmpty()) {
+                    // nobody is using this file so we close the channel
+                    channelRef.decRef();
+                    channelRef = null;
```

Comment:
Add `assert channelRef.recCount() == 0`?

Reply:
We can't assert that. We could still be using that specific channel in a background operation. This fact is to some degree at the heart of what motivated this PR: we don't want to lock here and wait for all the background fetchers/warmers etc. to be done with the file so we don't have to block and instead have the last of them to finish its work close it.

Reply:
Thanks, I wonder if we should then just change the comment a few lines up to not say "close" (or just remove the comment), at least that is what confused me.